### ISSUE TITLE
Plugin isn't working behind proxy

### ISSUE THREAD

**Issue by stese** [bordaigorl/sublime-evernote#42]
If you are using sublime text behind a company proxy, you basically have to set the proxy settings per module right to get this module working right now.

I had to set the company proxy for example for the plugin package control in the settings of the package control, because the sublime own proxy settings were ignored (see https://sublime.wbond.net/docs/settings#setting-http_proxy).

The possibility to set custom proxy settings for this module would be nice. For now, the plugin isn't working behind a proxy, even if the .bash_profile of the user or something similar is set up correctly.

**Comment by zwpaper** [comment]
it doesn't work...I have tried 3059 and 3047, neither of them work...looks like the same problem.

> Evernote plugin error: [WinError 10060] A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond
error: Evernote plugin error, please contact developer at
https://github.com/bordaigorl/sublime-evernote/issues

**Comment by tauren** [comment]
Too bad there is no easy fix to this. I will try it out at being my work proxy tomorrow, as I imagine it will affects me as well. I just got a pretty awesome evernote/sublime workflow going at home and was hoping to use it at work too.

**Comment by BattoorCjorn** [comment]
@tauren 
Jeah, I have the same problem, set up my workflow at home, and thought it was awsome, then I got at work and --> Doh! Not working :-1:

**Comment by tauren** [comment]
@BattoorCjorn So I just got in to work and tried it out. I'm not sure why, but everything seems to work properly and I didn't do anything special. I'm on OSX and have environment variables such as HTTP_PROXY configured properly.

**Comment by BattoorCjorn** [comment]
Yeah I know, it only occurs when the proxy has been configured within Internet Explorer, and not in the environment variables..., this is in a government managed network, so I'ts not up to me to fiddle with the environment variables. That is why I would like for applications to let us set a proxy override ...

**Comment by bordaigorl** [comment]
@BattoorCjorn just to check, what happens if you set the evironment vars locally?
Like in a Batch file with `set` and starting sublime?

**Comment by BattoorCjorn** [comment]
As it is in a "corporate environment", the only variables I can set are the _user defined environment variables_, not the _system defined variables_ <sup>(u need to be admin for that)</sup>.

So to answer the question, __nothing special__, and that is because  Sublime Text only looks at the _SYSTEM defined vars, instead of the SYSTEM + USER DEFINED_ vars ... 

This is the same problem I have with applications that use _PATH environment variables_ to locate certain _resources_ ... that never works, and I always need to set the PATH's to the resource files manually in the _setting file_ of _plugin_ itself.

**Comment by bordaigorl** [comment]
I see. Thanks for the clarification. Unfortunately I am not very experienced with this aspect.
Does anybody know if messing with `os.environ` from within ST can help?

**Comment by BattoorCjorn** [comment]
No, it won't help becouse the thrift (Phyton 3.x) library used in the evernote [API](https://github.com/evernote/evernote-sdk-python3), does not use these settings. 

The problem is that when they Ported thrift to 3.x, they left proxy support out of it. 


### Constructors THttpClient
Lets look at the [old 2.x implementation of the Thttpclient](https://github.com/evernote/evernote-sdk-python/blob/master/lib/thrift/transport/THttpClient.py):

```python
class THttpClient(TTransportBase):

    """Http implementation of TTransport base."""

    def __init__(
        self,
        uri_or_host,
        port=None,
        path=None,
        proxy_host=None,
        proxy_port=None
    ):
        """THttpClient supports two different types constructor parameters.
        THttpClient(host, port, path) - deprecated
        THttpClient(uri)
        Only the second supports https."""

        """THttpClient supports proxy
        THttpClient(host, port, path, proxy_host, proxy_port) - deprecated
        ThttpClient(uri, None, None, proxy_host, proxy_port)"""
```

So if we look at the constructor u can clearly see that in Python 2 support for proxy's was added, and used.

Now if u look at [the 3.x constructor](https://github.com/evernote/evernote-sdk-python3/blob/master/lib/thrift/transport/THttpClient.py) u notice that support has been teared down:

```python
class THttpClient(TTransportBase):
  """Http implementation of TTransport base."""

  def __init__(self, uri_or_host, port=None, path=None):
    """THttpClient supports two different types constructor parameters.
    THttpClient(host, port, path) - deprecated
    THttpClient(uri)
    Only the second supports https.
    """
```

So that means even if u would be able to set the environment `HTTP_PROXY` variable, you would __not be able to pass it to the constructor__ of the ThttpClient.

### Where is Thrift Python?
So now that we know what needs to be changed, __where do we change it__?

I notice [Thrift is compiled from this website](https://www.apache.org/dyn/closer.cgi?path=/thrift/0.9.2/thrift-0.9.2.exe), maybe just updating to the latest may already contain a fix?

**Comment by bordaigorl** [comment]
@BattoorCjorn yes, I noticed that as well but since THttpClient uses http.client I was wondering if there was a way to just set the http.client proxy setting globally.
in fact I did some experiments by manually porting the proxy related code of the new version to the Thrift shipped with the EvernoteSDK.
I cannot get past a "Message length exceeded" error. This error seems to be generated by the server in response to a request forwarded by the proxy (I am using a simple local proxy based on `BaseHTTPServer` for testing). At the moment I still do not know what is causing the problem.

As for updating Thrift, it would seem you need some sources we do not have from which Thrift generates the library....

**Comment by BattoorCjorn** [comment]
I have composed(stole :) ) [some code](https://gist.github.com/BattoorCjorn/70f07672d32e1ac3d323) that I will test tomorrow, hopefully its going in the right direction.

Was original Python 2.x using urllib2, adjusted it to work on 3.4.

**Comment by thorstenhuhn** [comment]
I would also love to see this problem fixed but I have a workaround to share:
Alternative A: I was able to reroute the ST plugins outgoing traffic transparently through the proxy using [https://www.proxifier.com/index.htm] for Mac which is also available for Windows and allows 30 days trial. Unfortunately 40 $ is pretty much money having only this one use case.
Alternative B (Mac only): The same thing works using [https://github.com/csujedihy/proximac] but I have to disable System Integrity Protection which leaves a bad taste, too.

**Comment by qunwang6** [comment]
anybody can fix this problem?
